File: 04_Binary_Tree/30_tree_path_finder/tree_path_finder.js
const pathFinder = (root, target) => {
  const result = pathFinderHelper(root, target);
  if (result === null) {
    return null;
  } else {
    return result.reverse();
  }
};

const pathFinderHelper = (root, target) => {
  if (root === null) return null;
  if (root.val === target) return [ root.val ];
  
  const leftPath = pathFinderHelper(root.left, target);
  if (leftPath !== null) {
    leftPath.push(root.val);
    return leftPath;
  }
  
  const rightPath = pathFinderHelper(root.right, target);
  if (rightPath !== null) {
    rightPath.push(root.val);
    return rightPath;
  }
  
  return null;
}

File: 04_Binary_Tree/30_tree_path_finder/tree_path_finder.py
# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.left = None
#     self.right = None

# def path_finder_slow(root, target):
#   if root is None:
#     return None
  
#   if root.val == target:
#     return [root.val]
  
#   left_path = path_finder_slow(root.left, target)
  
#   right_path = path_finder_slow(root.right, target)
  
#   if left_path is not None:
#     return [ root.val, *left_path]
  
#   if right_path is not None:
#     return [ root.val, *right_path]

def path_finder(root, target):
  result = _path_finder(root, target)
  if result is None:
    return None
  else:
    return result[::-1]

def _path_finder(root, target):
  if root is None:
    return None
  
  if root.val == target:
    return [root.val]
  
  left_path = path_finder_slow(root.left, target)
  right_path = path_finder_slow(root.right, target)

  if left_path is not None:
    left_path.append(root.val)
    return left_path

  if right_path is not None:
    right_path.append(root.val)
    return right_path